internal/pkg/cli/env_init.go

```diff
@@ -548,12 +555,18 @@ func (o *initEnvOpts) deployEnv(app *config.Application, customResourcesURLs map
 		Version:                  deploy.LatestEnvTemplateVersion,
 	}
 
+	if err := o.cleanUpDanglingRoles(o.appName, o.name); err != nil {
+		return err
+	}
 	if err := o.envDeployer.DeployAndRenderEnvironment(os.Stderr, deployEnvInput); err != nil {
 		var existsErr *cloudformation.ErrStackAlreadyExists
 		if errors.As(err, &existsErr) {
 			// Do nothing if the stack already exists.
 			return nil
 		}
+		// The stack failed to create due to an unexpect reason.
+		// Delete the retained roles created part of the stack.
+		o.tryDeletingEnvRoles(o.appName, o.name)
```

Comment:
If we delete the role before deleting the stack (since it failed to be deleted), will this make users not authorized to manage the env stack unless they manually create the env manager role again (if env upgrade happened before)?

Reply:
PH's comment makes me wonder whether it is guaranteed that if `deployAndRenderEnvironment` fails, then the env stack will be rolled back and delete 🤔

Reply:
Very good point! I think most of the time it will just be deleted but sometimes the rollback could fail (see `ROLLBACK_FAILED` in [here](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-describing-stacks.html)). It is just since we always try deleting the roles for a clean state when we create the environment so maybe it is ok to skip deleting the roles if the creation fails?

Reply:
I don't think I am fully following, so I'll try my best to clarify but please let me know if I'm not answering properly:

1. > If we delete the role before deleting the stack (since it failed to be deleted), will this make users not authorized to manage the env stack unless they manually create the env manager role again 

We don't pass a RoleARN for CloudFormation to assume when we initially create the stack. So on failure, it is safe to delete the roles that were retained since they were not used yet to mutate the cloudformation stack.

2. > `whether it is guaranteed that if deployAndRenderEnvironment fails, then the env stack will be rolled back and delete`

On failure, `deployAndRenderEnvironment` will indeed rollback back the stack but won't delete it. Instead, the second time that the customer runs `copilot env init` then on `Create` we will check if the stack requires deletion and if so first delete it and then re-create it: https://github.com/aws/copilot-cli/blob/1284849126a6a09b5ac25fe883501793d28f8a7f/internal/pkg/aws/cloudformation/cloudformation.go#L48-L65

3. > it is ok to skip deleting the roles if the creation fails

This would be true, if only the stack got deleted if the creation fails. Instead it goes into a `ROLLBACK_COMPLETE`  state I believe. So if the user just re-runs `copilot env init` and we don't clean up the IAM roles on failed create, then we will:
  1.  first check if the stack exists ✅, so skip 
  2. The stack tries to re-create the IAM but fails because the CFNExecRole already exists ❌
  
Does this help?

Reply:
Ah yes. Sorry I was stupid I wasn't aware it is creating an environment so env upgrade can't happen before.

Reply:
Yesss thanks for the explanation! It makes sense. I have no problem with the current implementation either but same question as Penghao's just to make sure my understanding aligns with the implementation.

Reply:
> env init will try to delete the existing IAM roles before deploying the stack right?

Yes that's right! We delete the roles iff:
1. There is no environment cloudformation stack
2. But there are IAM roles

This scenario means that `env delete` previously failed to delete the EnvManagerRole and it should be cleaned up before trying to re-create it. I updated the comment in the `cleanUpDanglingRoles` method, hopefully that explains a bit better this scenario.